Language/Classes/Instance_Methods/Method_noSuchMethod/forwarder_A01_t09.dart

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Let C be a concrete class, let L be the library that contains the
+/// declaration of C, and let m be a name. Then m is noSuchMethod forwarded in C
+/// iff one of the following is true:
+///
+/// • Requested in program: C has a non-trivial noSuchMethod, the interface of C
+/// contains a member signature S named m, and C has no concrete member named m
+/// and accessible to L that correctly overrides S (that is, no member named m
+/// is declared or inherited by C, or one is inherited, but it does not have the
+/// required signature). In this case we also say that S is noSuchMethod
+/// forwarded.
+///
+/// • Forced by privacy: There exists a direct or indirect superinterface D of C
+/// which is declared in a library L2 different from L, the interface of D
+/// contains a member signature S named m, m is a private name, and no
+/// superclass of C has a concrete member named m accessible to L2 that
+/// correctly overrides S. In this case we also say that S is noSuchMethod
+/// forwarded.
+///
+/// For a concrete class C, a noSuchMethod forwarder is implicitly induced for
+/// each member signature which is noSuchMethod forwarded.
+///
+/// @description Checks that for a concrete class, a `noSuchMethod` forwarder is
+/// invoked if `C` has no concrete member named `m`. Expect run-time error if a
+/// forwarder of a setter is called with a wrong type.
+/// @author sgrekhov22@gmail.com
+
+import "../../../../Utils/expect.dart";
+
+String log = "";
+
+abstract mixin class A {
+  void set a(String s);
+}
+
+class C1 {
+  void set m(String s);
+  noSuchMethod(Invocation i) {
+    log = "C1";
+  }
+}
+
+class C2 extends A {
+  dynamic noSuchMethod(Invocation i) {
+    log = "C2";
+  }
+}
+
+class C3 implements A {
+  dynamic noSuchMethod(Invocation i) {
+    log = "C3";
+  }
+}
+
+class C4 with A {
+  dynamic noSuchMethod(Invocation i) {
+    log = "C4";
+  }
+}
+
+mixin M1 {
+  void set m(String s);
+  noSuchMethod(Invocation i) {
+    log = "M1";
+  }
+}
+
+mixin M2 on A {
+  dynamic noSuchMethod(Invocation i) {
+    log = "M2";
+  }
+}
+
+mixin M3 implements A {
+  dynamic noSuchMethod(Invocation i) {
+    log = "M3";
+  }
+}
+
+class MA1 = Object with M1;
+class MA2 = A with M2;
+class MA3 = A with M3;
+
+enum E1 {
+  e1,
+  e2;
+
+  void set m(String s);
+  noSuchMethod(Invocation i) {
+    log = "E1";
+  }
+}
+
+enum E2 implements A {
+  e1,
+  e2;
+
+  dynamic noSuchMethod(Invocation i) {
+    log = "E2";
+  }
+}
+
+enum E3 with A {
+  e1,
+  e2;
+
+  dynamic noSuchMethod(Invocation i) {
+    log = "E3";
+  }
+}
+
+main() {
+  Expect.throws(() {C1().m = 1 as dynamic;});
```

Comment:
Thinking about this one more time, we probably have to use `(C1() as dynamic).m = 1` in order to be safe: It is possible that the compiler will simply generate `C1().m = (1 as dynamic) as String` based on the static knowledge about the interface of `C1`, and then we still won't know whether there's a forwarder or not. But if we do `(C1() as dynamic).m = 1` and get a type error at run time then there _must_ be a forwarder (and it does have a parameter type that works as expected).